File: leetcode/163-missing-ranges.cpp
class Solution
{
 public:
  vector<string>
  findMissingRanges(vector<int>& nums, int lower, int upper)
  {
    vector<string> ret;
    if (nums.empty() || nums.back() < upper)
      nums.push_back(upper+1);
    for (int i = 0, cur = lower; i < nums.size(); ++i) {
      string range;
      if (nums[i] > cur + 1)
        ret.push_back(to_string(cur) + "->" + to_string(nums[i]-1));
      else if (nums[i] > cur)
        ret.push_back(to_string(cur));
      cur = nums[i] + 1;
    }
    return ret;
  }
};

File: leetcode/163-missing-ranges.py
class Solution(object):
  def findMissingRanges(self, nums, lower, upper):
    """
    :type nums: List[int]
    :type lower: int
    :type upper: int
    :rtype: List[str]
    """
    ret = []
    pre = lower - 1
    start = lower
    for num in nums:
      if pre == num:
        continue
      pre = num
      if start != num:
        end = num - 1
        if start == end:
          ret += ['{0}'.format(start)]
        else:
          ret += ['{0}->{1}'.format(start, end)]
      start = num + 1
    if start < upper:
      ret += ['{0}->{1}'.format(start, upper)]
    elif start == upper:
      ret += ['{0}'.format(start)]
    return ret
